File: desktop/reencode.pl
#!/usr/bin/perl -w

# reencode.pl
#
# Encode all FLAC files in the current directory to MP3 with
# the specified bitrate.
#
# reencode.pl <bit rate (Kbps)>
#

use Term::ANSIColor;

# check for sane values
die "Usage: reencode.pl <bit rate (Kbps)>\n" if (! @ARGV);

$bit_rate = $ARGV[0];
die "Use a sane bit rate!\n" if ($bit_rate < 64);

@files = <*.flac>;

die "There are no FLAC files here...\n" if (scalar @files == 0);

print "Will encode ", scalar @files, " files to ${bit_rate}Kbps MP3 files. OK? [Y/n] ";
$answer = <STDIN>;
chomp $answer;

if ($answer eq "" || $answer eq "Y" || $answer eq "y") {

	# remove extension
	for ($i = 0; $i <= $#files; $i++) {
		$files[$i] =~ s/.flac$//;
	}

	# split up tasks
	$id = 0;
	while ($id < get_core_count()) {
#		push(@jobs_completed, 0);
		$pid = fork();

		if ($pid != 0) { # parent
			# assign the child a number and keep track of it
			push(@children, $pid);
			$id++;
		}
		else { # child
			last;
		}
	}

	if ($pid != 0) { # parent: should wait
		for ($i = 0; $i <= $#children; $i++) {
			waitpid($children[$i], 0);
			warn "Problems occured.\n" if ($? != 0);
		}
	}
	else { # child: do work
		if ($id > $#files) {
#			print "Not enough files; child #$id is bailing.\n";
			exit 0;
		}

		# convert files
		$status = 0;
		for ($i = $id; $i <= $#files; $i += get_core_count()) {

			# perform conversion
			$flac_file = $files[$i] . ".flac";
			$mp3_file = $files[$i] . ".mp3";
			system("ffmpeg -loglevel quiet -n -i \"$flac_file\" -ab $bit_rate \"$mp3_file\"");
			
			if ($? != 0) {
				print "$id: $flac_file -> $mp3_file ", colored("failed\n", "red\n");
				$status = 1;
			}
			else {
				print "$id: $flac_file -> $mp3_file ", colored("done\n", "green");
			}
		}

		exit $status;
	}
}
else {
	print "OK. Not doing anything then.\n";
	exit;
}

exit;

# count number of CPU cores on the system
sub get_core_count {
	if (`uname -a` =~ /Linux/) {
		($cc) = `cat /proc/cpuinfo` =~ /cpu cores\s+: (\d+)/;
	}
	else { # assume we can use sysctl
		$cc = `sysctl -n hw.ncpu`;
	}

	chomp $cc;

	return $cc;
}

File: desktop/reencode.py
#!/usr/bin/env python
"""
reencode.py

Encode all FLAC files in the current directory to MP3 with
the specified bitrate. This script performs multiple transocding
jobs at a time. The number of child processes it creates is equal
to the number of cores on the system.

reencode.py <bit rate (Kbps)>
"""
from argparse import ArgumentParser
from glob import glob
from multiprocessing import Pool
from subprocess import run, CalledProcessError
import os
import sys

def reencode(args):
    """Worker function to encode the file"""
    file_name, bit_rate = args
    cmd = ("ffmpeg -loglevel quiet -n -i \"{file_name}.flac\""
           " -ab {bit_rate}k \"{file_name}.mp3\"").format(
               bit_rate=bit_rate,
               file_name=file_name
           )
    try:
        run(cmd, shell=True, check=True)
    except CalledProcessError:
        print("failed to encode: ", file_name, file=sys.stderr)
        return 1
    else:
        print("success: ", file_name)
        return 0

def main():
    """main method"""
    parser = ArgumentParser()
    parser.add_argument("bit_rate", type=int)

    args = parser.parse_args()
    flac_files = glob("*.flac")

    if len(flac_files) == 0:
        print("There are no FLAC files here", file=sys.stderr)
        sys.exit(1)

    if args.bit_rate < 60:
        print("Use a sane bit rate!", file=sys.stderr)
        sys.exit(1)

    if not input("Will encode {num_files} files to {bit_rate}Kbps files. OK? [Y/n]".format(
            num_files=len(flac_files), bit_rate=args.bit_rate)).lower() in ["y", "yes"]:
        print("Ok, not doing anything then.")
        sys.exit(1)

    files = ((os.path.basename(i).split(".")[0], args.bit_rate) for i in flac_files)

    with Pool() as pool:
        sys.exit(sum(pool.map(reencode, files)))


if __name__ == "__main__":
    main()
